nucypher-core/src/dkg.rs

```diff
@@ -0,0 +1,120 @@
+use alloc::boxed::Box;
+use alloc::string::String;
+
+use serde::{Deserialize, Serialize};
+use umbral_pre::serde_bytes;
+
+use crate::conditions::{Conditions, Context};
+
+use crate::versioning::{
+    messagepack_deserialize, messagepack_serialize, ProtocolObject, ProtocolObjectInner,
+};
+
+/// The ferveo variant to use for the decryption share derivation.
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Copy, Clone)]
+pub enum FerveoVariant {
+    /// the simple variant requires n/n shares to decrypt
+    SIMPLE,
+    /// the precomputed variant requires m/n shares to decrypt
+    PRECOMPUTED,
+}
+
+/// A request for an Ursula to derive a decryption share.
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize)]
+pub struct ThresholdDecryptionRequest {
+    /// The ID of the ritual.
+    pub ritual_id: u16,
+    /// The ciphertext to generate a decryption share for.
+    #[serde(with = "serde_bytes::as_base64")]
+    pub ciphertext: Box<[u8]>,
+    /// A blob of bytes containing decryption conditions for this message.
+    pub conditions: Option<Conditions>,
+    /// A blob of bytes containing context required to evaluate conditions.
+    pub context: Option<Context>,
+    /// The ferveo variant to use for the decryption share derivation.
+    pub variant: FerveoVariant,
+}
+
+impl ThresholdDecryptionRequest {
+    /// Creates a new decryption request.
+    pub fn new(
+        ritual_id: u16,
+        ciphertext: &[u8],
+        conditions: Option<&Conditions>,
+        context: Option<&Context>,
+        variant: &FerveoVariant,
+    ) -> Self {
+        Self {
+            ritual_id,
+            ciphertext: ciphertext.to_vec().into(),
+            conditions: conditions.cloned(),
+            context: context.cloned(),
+            variant: *variant,
```

Comment:
@fjarri in response to your above comment about implementing `Copy`, it's appropriate to deference here?

Reply:
I think since it's small and Copy, you can just pass it by value in the method, so you won't need to dereference

Reply:
✔️